engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -1104,25 +1106,56 @@ static int CreateTextureAsync(lua_State* L)
     request->m_TextureResource   = (TextureResource*) resource;
     request->m_Handle            = request_handle;
     request->m_CallbackInfo      = callback_info;
-    request->m_Buffer            = upload_buffer;
-    request->m_UseUploadBuffer   = use_upload_buffer;
+    request->m_Buffer            = create_params.m_Buffer;
     request->m_PathHash          = create_params.m_PathHash;
+    request->m_RawData           = raw_data;
+    request->m_UseRawData        = use_raw_data;
 
     dmGraphics::TextureParams texture_params;
     texture_params.m_Width  = create_params.m_Width;
     texture_params.m_Height = create_params.m_Height;
     texture_params.m_Format = create_params.m_Format;
 
-    dmBuffer::GetBytes(request->m_Buffer, (void**)&texture_params.m_Data, &texture_params.m_DataSize);
-
     // If we are not using a specifically created upload buffer, we push the Lua object and increase its ref count,
     // this is so that we can make sure that the buffer to live during the upload.
-    if (!use_upload_buffer)
+    if (create_params.m_Buffer)
     {
         lua_pushvalue(L, 3);
         request->m_BufferRef = dmScript::Ref(L, LUA_REGISTRYINDEX);
     }
 
+    dmBuffer::GetBytes(request->m_Buffer, (void**) &texture_params.m_Data, &texture_params.m_DataSize);
+
+    // If the data is transcoded, we need an extra pass here to unpack the data before uploading it
+    if (is_transcoded)
```

Comment:
This block of code is essentially what the res_texture.cpp does. It would be nice if we could merge the code paths somehow since we eventually want to support installing custom decoders (like webp?), but it can perhaps wait when we do those refactors. Plus, this function is getting complex now :/